File: sdks/java/basys.sdk/src/main/java/org/eclipse/basyx/aas/backend/connected/aas/submodelelement/property/ConnectedCollectionProperty.java
package org.eclipse.basyx.aas.backend.connected.aas.submodelelement.property;

import java.util.Collection;

import org.eclipse.basyx.aas.api.exception.ServerException;
import org.eclipse.basyx.aas.api.exception.TypeMismatchException;
import org.eclipse.basyx.aas.api.metamodel.aas.submodelelement.property.ICollectionProperty;
import org.eclipse.basyx.aas.api.metamodel.aas.submodelelement.property.PropertyType;
import org.eclipse.basyx.aas.impl.metamodel.hashmap.aas.submodelelement.property.Property;
import org.eclipse.basyx.vab.core.proxy.VABElementProxy;

/**
 * Connects to a PropertySingleValued as specified by DAAS containing a
 * collection
 * 
 * @author schnicke
 *
 */
public class ConnectedCollectionProperty extends ConnectedProperty implements ICollectionProperty {

	public ConnectedCollectionProperty(VABElementProxy proxy) {
		super(PropertyType.Collection, proxy);
	}

	@Override
	public void set(Collection<Object> collection) throws ServerException {
		try {
			getProxy().setModelPropertyValue(Property.VALUE, collection);
		} catch (Exception e) {
			throw new ServerException(e.getClass().toString(), e.getMessage());
		}
	}

	@Override
	public void add(Object newValue) throws ServerException, TypeMismatchException {
		try {
			getProxy().createValue(Property.VALUE, newValue);
		} catch (Exception e) {
			throw new ServerException(e.getClass().toString(), e.getMessage());
		}
	}

	@Override
	public void remove(Object objectRef) throws ServerException {
		try {
			getProxy().deleteValue(Property.VALUE, objectRef);
		} catch (Exception e) {
			throw new ServerException(e.getClass().toString(), e.getMessage());
		}
	}

	@Override
	public Collection<Object> getElements() throws ServerException {
		return getCollection();
	}

	@Override
	public int getElementCount() throws ServerException {
		return getCollection().size();
	}

	@SuppressWarnings("unchecked")
	private Collection<Object> getCollection() {
		return (Collection<Object>) getProxy().getModelPropertyValue(Property.VALUE);
	}
}

File: sdks/c++/basys.sdk.cc/src/aas/backend/connected/aas/submodelelement/property/ConnectedCollectionProperty.cpp
/*
 * ConnectedCollectionProperty.cpp
 *
 *      Author: wendel
 */

#include "ConnectedCollectionProperty.h"
#include "aas/submodelelement/property/IProperty.h"

namespace basyx {
namespace aas {
namespace backend {
namespace connected {

using namespace submodelelement::property;

ConnectedCollectionProperty::ConnectedCollectionProperty(std::shared_ptr<vab::core::proxy::IVABElementProxy> proxy) :
  ConnectedProperty(PropertyType::Collection, proxy)
{}

void ConnectedCollectionProperty::set(const basyx::objectCollection_t & collection) const
{
  this->getProxy()->updateElementValue(PropertyPaths::VALUE, collection);
}

void ConnectedCollectionProperty::add(const basyx::any & newValue)
{
  this->getProxy()->createElement(PropertyPaths::VALUE, newValue);
}

void ConnectedCollectionProperty::remove(const basyx::any & objectRef)
{
  this->getProxy()->deleteElement(PropertyPaths::VALUE, objectRef);
}

basyx::objectCollection_t ConnectedCollectionProperty::getElements() const
{
  return this->retrieveObject().Get<basyx::objectCollection_t>();
}

int ConnectedCollectionProperty::getElementCount() const
{
  return this->retrieveObject().Get<basyx::objectCollection_t>().size();
}

}
}
}
}